algebird-test/src/main/scala/com/twitter/algebird/ApproximateProperty.scala

```diff
@@ -0,0 +1,96 @@
+package com.twitter.algebird
+
+import org.scalacheck.{ Gen, Prop }
+import org.scalacheck.util.Pretty
+
+trait ApproximateProperty {
+  type Exact
+  type Approx
+  type Input
+  type Result
+
+  def exactGenerator: Gen[Exact]
+  def inputGenerator(e: Exact): Gen[Input]
+
+  def makeApproximate(e: Exact): Approx
+
+  def exactResult(e: Exact, i: Input): Result
+  def approximateResult(a: Approx, i: Input): Approximate[Result]
+}
+
+object ApproximateProperty {
+
+  /**
+   *  Generates a list of exactly n Ts.
+   *  Useful because `Gen.listOfN(n, gen).sample` gives us Option[List[T]],
+   *  while we often want List[T].
+   */
+  private def genListOf[T](n: Int, gen: Gen[T]): List[T] = {
+    Gen.listOfN(n, gen).sample match {
+      case Some(xs) => xs
+      case _ => genListOf(n, gen)
+    }
+  }
+
+  private def successesAndProbabilities(a: ApproximateProperty, objectReps: Int, inputReps: Int): (Int, Double, List[String]) = {
+    val list = genListOf(objectReps, a.exactGenerator)
+      .flatMap { exact =>
+        val approx = a.makeApproximate(exact)
+        genListOf(inputReps, a.inputGenerator(exact)).map { input =>
+          val approxResult = a.approximateResult(approx, input)
+          val exactResult = a.exactResult(exact, input)
+
+          val success = approxResult.boundsContain(exactResult)
+          val successInt = if (success) 1 else 0
+          // val messages = if (success) List() else List(s"Exact: $exact. Approx: $approx. Exact result: $exactResult. Approx result: $approxResult.")
+          val messages = if (success) List() else List(s"Exact result: $exactResult. Approx result: $approxResult.")
+          (successInt, approxResult.probWithinBounds, messages)
+        }
+      }
+    val monoid = implicitly[Monoid[(Int, Double, List[String])]]
+    monoid.sum(list)
+  }
+
+  def toProp(a: ApproximateProperty, objectReps: Int, inputReps: Int, falsePositiveRate: Double): Prop =
+    new Prop {
+      def apply(params: Gen.Parameters) = {
+        require(0 <= falsePositiveRate && falsePositiveRate <= 1)
+
+        val (successes, sumOfProbabilities, exacts) = successesAndProbabilities(a, objectReps, inputReps)
+        val n = objectReps * inputReps
+
+        // Computed from Hoeffding's inequality, might be inaccurate
+        // TODO Make sure this is correct
```

Comment:
The general case (https://en.wikipedia.org/wiki/Hoeffding%27s_inequality#General_case) is
```
P(sum(X) - sum(E[X]) >= nt) <= exp(-2*n*t^2)
```
(here I used `sum(X) = X_1 + ... + X_n` instead of `Xbar = 1/n (X_1 + ... + X_N)`)
So we have
```
   exp(-2*n*t^2) = fprate
=> t = sqrt(-log p / (2*n) )
```